### DISCUSSION TITLE
Calling (a bunch of) Javascript Functions from a CefSharp Application

### DISCUSSION THREAD

**Discussion by jl2** [cefsharp/CefSharp#4377] [Q&A]
We're using CefSharp to embed a page with a WebGL canvas in our application, allowing us to use the same renderer for both our desktop and web applications.

It works remarkably well, but in some situations we're calling a large number of Javascript functions to query the geometry loaded in the WebGL canvas and Javascript context.  This was cheap and fast with our old renderer, which used C++/CLI and OpenGL.

Profiling in Visual Studio shows >99% of the C# time is spent waiting for `EvaluateScriptAsync`, and profiling in the Cef dev tools shows that >70% of that time is spent in "Compile Script" and "Compile Code".

I'm looking at ways to reduce that time, and one option I found is to use the `CefV8Context`'s [`GetGlobal`](https://cef-builds.spotifycdn.com/docs/108.4/classCefV8Context.html#a2daf239008f5bf22d5aaaef8658847d1) method, and then use [`ExecuteFunction`](https://cef-builds.spotifycdn.com/docs/108.4/classCefV8Value.html#a2a578d33aa43a7e446867cc0b256aa04) to call the functions we need.

My understanding is that going that route would avoid the compilation, and so it should execute faster, but it doesn't look like those capabilities of the `CefV8Context` are exposed in CefSharp.

So my questions are:

Is there a way to execute Javascript functions in CefSharp directly via the `CefV8Context`, without compiling Javascript strings?

If not, any suggestions for speeding up a large number of Javascript calls?  The alternatives I've come up with involve a pretty big refactor for our existing code...

**Comment by amaitland** [comment]
> I'm looking at ways to reduce that time, and one option I found is to use the `CefV8Context`'s [`GetGlobal`](https://cef-builds.spotifycdn.com/docs/108.4/classCefV8Context.html#a2daf239008f5bf22d5aaaef8658847d1) method, and then use [`ExecuteFunction`](https://cef-builds.spotifycdn.com/docs/108.4/classCefV8Value.html#a2a578d33aa43a7e446867cc0b256aa04) to call the functions we need.

You cannot directly call `ExecuteFunction` on the `Global CefV8Value`, you need to obtain a `CefV8Value` that represents a function first. 

`ExecuteFunction` is exposed in the form of [IJavascriptCallback](https://cefsharp.github.io/api/109.1.x/html/T_CefSharp_IJavascriptCallback.htm).

You can execute javascript that returns a function, then execute that function multiple times via [IJavascriptCallback](https://cefsharp.github.io/api/109.1.x/html/T_CefSharp_IJavascriptCallback.htm). If your planning on executing the callback multiple times then don't forget to check the `CanExecute` property.

See https://github.com/cefsharp/CefSharp/blob/cefsharp/109/CefSharp.Test/Javascript/JavascriptCallbackFacts.cs for some test case examples.

**Reply by jl2** [reply]
Thank you for taking the time to look at this!

This looks like exactly what I need.  Our code will look something like this:
```
// Return the JS function
var javascriptResponse = await browser.EvaluateScriptAsync("window.ourDocument.myQueryFunction");
Assert.True(javascriptResponse.Success);

// Convert to a callback
var myJsFunction = (IJavascriptCallback)javascriptResponse.Result;

// Call the function for each ID
foreach (var id in idList) {
    var callbackResponse = await myJsFunction.ExecuteAsync(queryValue);
    // Process result
}
```

**Reply by aimcosoftware** [reply]
I'd be really interested to see if you get any speed improvements, as I'm always looking for tips on this ;-)

I've just done a very quick test which seems to suggest the additional call to the JS function is actually making it slower.

My understanding is that the JS execution loop is the limiting factor for multiple calls, as it gives priority to display and responsiveness.

**Reply by jl2** [reply]
So far this seems to give a good speed up for us.

What I ended up doing is creating a dictionary of function names to `IJavascriptCallback` objects, and the first time a function is called I build the anonymous wrapper, execute it with `EvaluateScriptAsync` to get an `IJavascriptCallback`, and then add that callback to the map.  From then on I run it with `functionMap[functionName].ExecuteAsync(argumentArray)`.

We might have to optimize our Javascript sooner or later, or cache things in the C# code, but for now those changes were enough to get rid of almost all of time spent in "Compile Script" in the hotpath.

One issue I ran into immediately was that I needed to wrap the Javascript functions from `window.myControl.ourDocument.myFunction` in an anonymous function block: `(function (arg1, arg2) { window.myControl.ourDocument.myFunction(arg1, arg2); })` to make "this" behave correctly when the callback runs.  That could be some kind of Javascript misunderstanding on my part, but the anonymous wrapper fixed it for now.

**Reply by aimcosoftware** [reply]
Thanks for posting, tips on speeding up JS access from the wrapper are always useful ;-)

**Reply by amaitland** [reply]
> One issue I ran into immediately was that I needed to wrap the Javascript functions from window.myControl.ourDocument.myFunction in an anonymous function block: (function (arg1, arg2) { window.myControl.ourDocument.myFunction(arg1, arg2); }) to make "this" behave correctly when the callback runs.

We don't specify a `this` object when executing a callback so the [globalThis](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis) is used. This is something I'd like to improve, unfortunately in the case where [CefV8Context::Eval](https://cef-builds.spotifycdn.com/docs/109.1/classCefV8Context.html#a160b647d434ed0bcd8518c0042e7d991) is used to directly obtain a function I've been unable to find a means of obtaining the parent object. [Function.prototype.caller](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/caller) is deprecated (didn't work when I tried in newer `Chromium` either). Open to suggestions if someone has one.

You should be able to use a standard [bind](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind) when returning the function to specify the `this`. Something like:

```js
window.myControl.ourDocument.myFunction.bind(window.myControl.ourDocument);
```

**Reply by aimcosoftware** [reply]
`Function.prototype.caller` is still working in Chrome 109.0.5414.120

Can you use `window.myControl.ourDocument.myFunction.call(myObj, args)`

**Reply by amaitland** [reply]
> We don't specify a `this` object when executing a callback so the [globalThis](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis) is used. This is something I'd like to improve, unfortunately in the case where [CefV8Context::Eval](https://cef-builds.spotifycdn.com/docs/109.1/classCefV8Context.html#a160b647d434ed0bcd8518c0042e7d991) is used to directly obtain a function I've been unable to find a means of obtaining the parent object.

```js
const module = {
  x: 42,
  getX: function() {
    return this.x;
  }
};
module.getX;
```

Using the above as an example, we get back a `CefV8Value` where `CefV8Value::IsFunction() == true`. Using the `CEF API` how to obtain the correct `this` value (a reference to `module`)?

- [CEF V8 Unit Tests](https://github.com/chromiumembedded/cef/blob/36ee304ed48f3ffbeef9edf4fa7762d3837af9d9/tests/ceftests/v8_unittest.cc), the only example I see that passes in an explicit `this`, a new object is created, so not sure that would be an improvement in behaviour.
- [v8::value](https://source.chromium.org/chromium/chromium/src/+/main:v8/include/v8-value.h;l=30?q=v8::Value&ss=chromium) API in `Chromium`, nothing I could see that would achieve this.

I'll ask on CEF forum, see if any other CEF users have ideas.

For now it'll have to remain up to the end user to explicitly specify a `this` value using `bind/call/apply`.

**Comment by aimcosoftware** [comment]
Don't know if this helps for your case, but one of the things I do to overcome this, is to move your results into javascript, then return an object (either as expando or JSON)
It does take some refactoring, but hopefully, not a lot. It allows you to do it the slower way to test, then move into JS to speed it up.
e.g.
```
a =  await ExecuteAsync("return a")
b =  await ExecuteAsync("return b")

```
```
r = await ExecuteAsync("obj.a=a;obj.b=b;return obj")
a = r.a
b = r.b
```

**Reply by jl2** [reply]
Thank you for taking the time to help with this!

This is similar to a solution we had in mind - basically collecting small snippets into larger scripts and running them with a single call to `EvaluateScriptAsync`.

I'm going to try the solution from amaitland first, but this is our backup plan.
